Adopt more smart pointers in RemoteCDMInstanceProxy and RemoteVideoFrameObjectHeap

RupinMittal opened the pull request:
#### b5c6650e0ec320d8cb4ce92e14d45266cc65ca87
<pre>
Adopt more smart pointers in RemoteCDMInstanceProxy and RemoteVideoFrameObjectHeap
<a href="https://bugs.webkit.org/show_bug.cgi?id=279785">https://bugs.webkit.org/show_bug.cgi?id=279785</a>
<a href="https://rdar.apple.com/136103868">rdar://136103868</a>

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/GPUProcess/media/RemoteCDMInstanceProxy.cpp:
(WebKit::RemoteCDMInstanceProxy::unrequestedInitializationDataReceived):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::remoteVideoFrameObjectHeapQueueSingleton):
(WebKit::RemoteVideoFrameObjectHeap::create):
(WebKit::RemoteVideoFrameObjectHeap::releaseVideoFrame):
(WebKit::RemoteVideoFrameObjectHeap::getVideoFrameBuffer):
(WebKit::RemoteVideoFrameObjectHeap::pixelBuffer):
(WebKit::remoteVideoFrameObjectHeapQueue): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxy.h:
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidClose):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage):

Canonical link: <a href="https://commits.webkit.org/283775@main">https://commits.webkit.org/283775@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c643120542f2b84c9ea7f99ad2133f4c4ece72a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67225 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19857 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71276 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18358 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54402 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18152 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53882 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12333 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70292 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58165 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34416 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39510 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61474 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15899 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72963 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11184 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15238 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61358 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11217 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61437 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9178 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2772 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10222 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42409 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43486 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44672 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43227 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
```diff
@@ -150,18 +150,18 @@ void RemoteVideoFrameObjectHeapProxyProcessor::newConvertedVideoFrameBuffer(std:
-        [&](auto& semaphore) { connection.send(Messages::RemoteVideoFrameObjectHeap::SetSharedVideoFrameSemaphore { semaphore }, 0); },
-        [&](auto&& handle) { connection.send(Messages::RemoteVideoFrameObjectHeap::SetSharedVideoFrameMemory { WTFMove(handle) }, 0); });
+        [&](auto& semaphore) { connection->send(Messages::RemoteVideoFrameObjectHeap::SetSharedVideoFrameSemaphore { semaphore }, 0); },
+        [&](auto&& handle) { connection->send(Messages::RemoteVideoFrameObjectHeap::SetSharedVideoFrameMemory { WTFMove(handle) }, 0); });
```
Locally, the style-checker complained about this line saying "More than one command on the same line". Is this ok? If not, how I should fix it?

rreno replied:
This is more or less the same as a `WTF::switchOn`. I'd call it an edge case the style checker isn't able to recognize.

rreno replied:
*more or less the same in terms of formatting.

RupinMittal replied:
Sounds good--I'll ignore the warning from the style checker then. Thank you!

rreno commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
```diff
@@ -47,10 +47,9 @@ namespace WebKit {
 {
-    static NeverDestroyed queue = WorkQueue::create("org.webkit.RemoteVideoFrameObjectHeap"_s, WorkQueue::QOS::UserInteractive);
-    return queue.get();
+    return WorkQueue::create("org.webkit.RemoteVideoFrameObjectHeap"_s, WorkQueue::QOS::UserInteractive);
```
This is a pretty major change to the lifecycle management of this object! I think to fix the static analyzer issue here you want to rename this to some flavor of `fooSingleton()` and keep it a NeverDestroyed.

RupinMittal replied:
Oops! You're totally right--I missed the NeverDestroyed there. Since the analyzer was compiling about `remoteVideoFrameObjectHeapQueue()` being used as a call argument but being uncounted, I changed its result to be held as a ref in `RemoteVideoFrameObjectHeap::create`. I believe that should fix the static analyzer issue. Should I still rename the function to include `singleton` or does it look ok now?
Analyzer issue: https://build.webkit.org/results/Apple-Sonoma-Smart-Pointer-Static-Analyzer-Build/283707@main%20(1199)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-RemoteVideoFrameObjectHeap.cpp-unknown-0-8035e9.html#EndPath

rreno replied:
I think if you keep everything else the same but rename the getter to be `remoteVideoFrameObjectHeapQueueSingleton()` it will be OK with you returning a raw reference. At least that's how I understood rniwa's message on webkit-dev the other day.

rreno replied:
I think a `Ref` is redundant since the object is static and can never be destroyed.

RupinMittal replied:
Oh I see! That's pretty neat. I didn't think the analyzer would examine function names too--just return types. Thank you for letting me know. I've fixed it in the latest iteration. Is webkit-dev one of our slack channels? I'm not sure if I'm in it.

rreno replied:
It's the public mailing list. You can join here: https://lists.webkit.org/mailman/listinfo/webkit-dev

RupinMittal replied:
Got it--thank you.

cdumez commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
```diff
@@ -56,7 +56,8 @@ static WorkQueue& remoteVideoFrameObjectHeapQueue()
 {
     auto heap = adoptRef(*new RemoteVideoFrameObjectHeap(WTFMove(connection)));
-    heap->m_connection->addWorkQueueMessageReceiver(Messages::RemoteVideoFrameObjectHeap::messageReceiverName(), remoteVideoFrameObjectHeapQueue(), heap);
+    Ref workQueue { remoteVideoFrameObjectHeapQueue() };
```
We should not introduce a local variable like this.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
```diff
@@ -56,7 +56,8 @@ static WorkQueue& remoteVideoFrameObjectHeapQueue()
     auto heap = adoptRef(*new RemoteVideoFrameObjectHeap(WTFMove(connection)));
-    heap->m_connection->addWorkQueueMessageReceiver(Messages::RemoteVideoFrameObjectHeap::messageReceiverName(), remoteVideoFrameObjectHeapQueue(), heap);
+    Ref workQueue { remoteVideoFrameObjectHeapQueue() };
+    heap->m_connection->addWorkQueueMessageReceiver(Messages::RemoteVideoFrameObjectHeap::messageReceiverName(), workQueue, heap);
```
Check with @rniwa but I think maybe we just need to add `Singleton` to this function name to silence the warning:
```
remoteVideoFrameObjectHeapQueue()
```

rniwa replied:
Yeah, we can rename `remoteVideoFrameObjectHeapQueue` to `remoteVideoFrameObjectHeapQueueSingleton` and the warning should go away.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
```diff
@@ -150,18 +150,18 @@ void RemoteVideoFrameObjectHeapProxyProcessor::newConvertedVideoFrameBuffer(std:
 RefPtr<NativeImage> RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage(const WebCore::VideoFrame& videoFrame)
 {
-    auto& connection = WebProcess::singleton().ensureGPUProcessConnection().connection();
+    Ref connection = WebProcess::singleton().ensureGPUProcessConnection().protectedConnection();
```
You don't need BOTH `Ref` and `protectedConnection()`. Just this is enough:
```
Ref connection = WebProcess::singleton().ensureGPUProcessConnection().connection();
```

RupinMittal replied:
Good point-fixed!

RupinMittal committed:
Adopt more smart pointers in RemoteCDMInstanceProxy and RemoteVideoFrameObjectHeap
https://bugs.webkit.org/show_bug.cgi?id=279785
rdar://136103868

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/GPUProcess/media/RemoteCDMInstanceProxy.cpp:
(WebKit::RemoteCDMInstanceProxy::unrequestedInitializationDataReceived):
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
(WebKit::remoteVideoFrameObjectHeapQueueSingleton):
(WebKit::RemoteVideoFrameObjectHeap::create):
(WebKit::RemoteVideoFrameObjectHeap::releaseVideoFrame):
(WebKit::RemoteVideoFrameObjectHeap::getVideoFrameBuffer):
(WebKit::RemoteVideoFrameObjectHeap::pixelBuffer):
(WebKit::remoteVideoFrameObjectHeapQueue): Deleted.
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxy.h:
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.cpp:
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::gpuProcessConnectionDidClose):
(WebKit::RemoteVideoFrameObjectHeapProxyProcessor::getNativeImage):

Canonical link: https://commits.webkit.org/283775@main
